File: solutions/Leetcode_653/leetcode_653.cpp
#include <stack>
#include <unordered_set>

class Solution1
{
public:
    bool findTarget(TreeNode* root, int k)
    {
        std::unordered_set<int> set;
        std::stack<TreeNode*> s;
        s.push(root);
        while (!s.empty())
        {
            TreeNode* node = s.top();
            s.pop();
            if (node != nullptr)
            {
                int target = k - node->val;
                if (set.find(target) != set.end())
                    return true;
                set.insert(node->val);
                s.push(node->left);
                s.push(node->right);
            }
        }
        return false;
    }
};


class Solution2
{
TreeNode* _root = nullptr;

public:
    bool findTarget(TreeNode* root, int k)
    {
        if (this->_root == nullptr)
            this->_root = root;
        if (root == nullptr)
            return false;
        return (_dfs(root, k - root->val) || findTarget(root->left, k) || findTarget(root->right, k));
    }

private:
    bool _dfs(TreeNode* node, int target)
    {
        TreeNode* root = this->_root;
        while (root != nullptr)
        {
            if (root->val == target && root != node)
                return true;
            if (root->val > target)
                root = root->left;
            else
                root = root->right;
        }
        return false;
    }
};

File: solutions/Leetcode_653/leetcode_653.py
from typing import Optional

from leetcode.python.tree import TreeNode


class Solution1:
    def find_target(self, root: Optional[TreeNode], k: int) -> bool:
        map = {}
        stack = [root]
        while stack:
            node = stack.pop()
            if node:
                target = k - node.val
                if map.get(target, False):
                    return True
                map[node.val] = True
                stack.append(node.left)
                stack.append(node.right)
        return False


class Solution2:
    _root = None

    def find_target(self, root, k):
        if not Solution2._root:
            Solution2._root = root
        if not root:
            return False
        return self._dfs(root, k-root.val) or self.find_target(root.left, k) or self.find_target(root.right, k)

    def _dfs(self, node, k):
        root = Solution2._root
        while root:
            if root.val == k and root != node:
                return True
            if root.val > k:
                root = root.left
            else:
                root = root.right
        return False
